### PR TITLE
Fire action if order line item has been restored

### PR DESCRIPTION
**Pull request by danielvonmitschke**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).

### Changes proposed in this Pull Request:

There is an existing action `woocommerce_reduce_order_item_stock` in the `wc_reduce_stock_levels()` function.
The corresponding `wc_increase_stock_levels()` function does not have such an action.
It only has a more general action after all items have been restored, but it is not possible to know which items have been restored.
This new action will trigger for each restored line item.

### How to test the changes in this Pull Request:

You can test as follows:
```
add_action( 'woocommerce_restore_order_item_stock', function($item, $new_stock, $old_stock, $order) {
    $product  = $item->get_product();
    $item_name = $product->get_formatted_name();
    error_log('Stock level of ' . $item_name . ' has been increased from ' . $old_stock . ' to ' . $new_stock . ' from order ' . $order->get_id());
}, 10, 4);
```
Then cancel an existing order that previously had the stock reduced (e.g. with status on-hold or processing) and check your log file :)

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Add action hook 'woocommerce_restore_order_item_stock' triggered when order line items are restored.

</details>

### CODE REVIEW

**Review by jorgeatorres — CHANGES_REQUESTED** [review]
Hi @danielvonmitschke!

Thanks for your contribution. Overall, this looks good to me but I wonder if we should make this new hook align with `woocommerce_reduce_order_item_stock` ([see in code](https://github.com/woocommerce/woocommerce/blob/92f94248cd3e7e01e4c49738acd35a8beeccd785/plugins/woocommerce/includes/wc-stock-functions.php#L215)) which is triggered when the opposite thing happens: an order item is reduced.

I wonder if we can implement a similar `$change` array so that the two actions can be used similarly. As I said, there's nothing wrong with your approach and it is, in fact, more direct than sending an array per item to the hook, but I think consistency would be best in this case.

What do you think?

**Comment by danielvonmitschke** [comment]
Hi @jorgeatorres,

thank you for the suggestions.
Let me have a look in the coming days and I will get back to you.

Best,

daniel

**Comment by danielvonmitschke** [comment]
Dear @jorgeatorres,

sorry for the late reply. I was involved in another large project.

First of all let me issue a slight correction to my description above.
This hook does NOT fire on refunded items.
It only fires on cancelled (on-hold) orders where previous reduced stock is being restored.
For refunded order items there already is a separate hook called `woocommerce_restock_refunded_item`.

Regarding your suggestion you are right on the one hand (having the one hook and its opposite behave the same).
But I would have two other arguments against it:
1. This would require to change the code quite a bit - I wanted to be this as less an impact as possible
2. The other corresponding hook `woocommerce_restock_refunded_item` ([see in code](https://github.com/woocommerce/woocommerce/blob/92f94248cd3e7e01e4c49738acd35a8beeccd785/plugins/woocommerce/includes/wc-order-functions.php#L784)) behaves pretty much like the one I introduced in my change request. 
So I would rather have these two behave similar.

Let me know what you think.

Best, Daniel

**Comment by jorgeatorres** [comment]
Hey @danielvonmitschke! 👋 

Thanks for getting back to us. All of these functions do show some inconsistencies, so I believe your approach should be fine too. That said, before we can merge this, the docblock for the new hook should be updated to include a `@since` tag (which should point to 9.1.0, assuming we can merge before the code freeze) and apparently there were some other checks failing but GitHub no longer reports the details.

Could you please rebase your changes off the latest `trunk` so we get a more up-to-date to run CI on?

Thanks!

**Comment by danielvonmitschke** [comment]
@jorgeatorres I added the @since tag and also rebased off the latest trunk.
I don't know much about CI actions, so please let me leave this up to you. :)

**Review by jorgeatorres — CHANGES_REQUESTED** [review]
Hey @danielvonmitschke!

Thanks again. Things are looking good to me, but we still need to address a couple of tiny things:

- In `wc-stock-functions.php` you seem to have used spaces for indentation instead of tabs, which is what our coding standards expect. As such, the linting checks are not passing: https://github.com/woocommerce/woocommerce/actions/runs/9520557412/job/26315576458?pr=40848.
- In your example code (testing instructions) it's important to indicate that the order must be in a specific status and not just "any order". Also, the PHP code needs an adjustment: the callback you've specified in `add_action()` receives 4 parameters, yet you didn't pass this number of args (`$accepted_args`) to `add_action()`.

Other than that, this is pretty much ready to go. Please let me know what you think.

**Comment by danielvonmitschke** [comment]
Dear @jorgeatorres,
I took care of the indents and also updated the testing instructions and corresponding php code.
Let me know if there is anything else that needs to be addressed.

**Review by jorgeatorres — APPROVED** [review]
LGTM. Thank you @danielvonmitschke for making the updates!

**Pull request merged**
